File: leetcode/python/144. 二叉树的前序遍历.py
#!/bin/bash python
"""
144. 二叉树的前序遍历

两种思路：递归法，迭代法
"""

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ret = []
        def front(root):
            if root == None:
                return
            ret.append(root.val)
            front(root.left)
            front(root.right)
        front(root)
        return ret
    
    def preorderTraversal2(self, root):
        """
        堆栈的方式
        """
        ret = []
        stack = [root] if root else []

        while stack:
            node = stack.pop()
            ret.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        
        return ret
	
    def preorderTraversal3(self, root):
        """
        更加标准的堆栈形式
        """
        ret, stack, curr = [], [], root
        while stack or curr:
            if curr:
                ret.append(curr.val) # 遍历根节点
                stack.append(curr)
                curr = curr.left # 遍历左子树
            else:
                curr = stack.pop().right # 遍历右子树
        return ret

File: leetcode/cpp/144. 二叉树的前序遍历.cpp
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // 先序遍历
    void front(TreeNode* root, vector<int>& ret) {
        if (root == nullptr) {
            return;
        }
        ret.push_back(root->val);
        front(root->left);
        front(root->right);
        return;
    }

    // 递归法
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret;
        front(root, ret);
        return ret;
    }

    // 迭代法
    vector<int> preorderTraversal2(TreeNode* root) {
        std::vector<int> ret;
        std::stack<TreeNode*> stack;
        if (root) {
            stack.push(root);
        }

        while (!stack.empty()) {
            TreeNode* node = stack.top();
            stack.pop();
            ret.push_back(node->val);
            if (node->right) {
                stack.push(node->right);
            }
            if (node->left) {
                stack.push(node->left);
            }
        }

        return ret;
    }

	// 更加标准的迭代方式
	vector<int> preorderTraversal3(TreeNode* root) {
		std::vector<int> ret;
		std::stack<TreeNode*> stack;
		TreeNode* curr = root;
		while (!stack.empty() || curr) {
			if (curr) {
				ret.push_back(curr->val); // 遍历根节点
				stack.push(curr);
				curr = curr->left; // 遍历左子树
			} else {
				curr = stack.top()->right; // 遍历右子树
				stack.pop();
			}
		}
		return ret;
	}
};